recipes/defiant/build.sh

```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+export C_INCLUDE_PATH=${C_INCLUDE_PATH}:${PREFIX}/include
+export LIBRARY_PATH="${PREFIX}/lib"
+export LD_LIBRARY_PATH="${PREFIX}/lib"
+
+mkdir -p ${PREFIX}/bin
+
+unzip defiant.zip
+
+# Fix defiant.c for OSX
+if [ `uname` == Darwin ]; then
+    sed -i.bak 's/restrict argv\[\]/argv\[\]/g' defiant.c
```

Comment:
This looks weird. I’ll have to look at the code for context.

Reply:
If I run the `gcc` command as it is, I am getting the following error:
**$ ${CC} -O4 -o defiant defiant.c -Wall -pedantic -std=gnu11 -lm -fopenmp -I${PREFIX}/include**
clang-4.0: warning: -O4 is equivalent to -O3 [-Wdeprecated]
In file included from defiant.c:12:
./read_bisulfite_data.h:99:25: warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
        if ((number_of_columns == 1)) {// && (strstr(temporary_char_string,"chr") != NULL)) {
             ~~~~~~~~~~~~~~~~~~^~~~
./read_bisulfite_data.h:99:25: note: remove extraneous parentheses around the comparison to silence this warning
        if ((number_of_columns == 1)) {// && (strstr(temporary_char_string,"chr") != NULL)) {
            ~                  ^   ~
./read_bisulfite_data.h:99:25: note: use '=' to turn this equality comparison into an assignment
        if ((number_of_columns == 1)) {// && (strstr(temporary_char_string,"chr") != NULL)) {
                               ^~
                               =
defiant.c:36:5: error: second parameter of 'main' (argument array) must be of type 'char **'
int main (int argc, char *restrict argv[]) {
    ^
1 warning and 1 error generated.

I tried few things, but removing `restrict` from the definition of `main` worked. Please let me know if there are better ways of fixing this error.

Reply:
No, that's correct.

Background:
`restrict` tells the compiler that the object a pointer references will not be referenced by any other pointer. That's relevant e.g. for something that copies data. E.g., if you do:
```
void copy(OBJ** source, OBJ** dest, int n) {
for (int i = 0; i <n; ++i) {
  dest[i] = source[i];
}
```
then the compiler can't know whether the regions overlap. You could be copying data within the same array, so the loop has to be executed in order. If you add the `restrict` keyword, it means that `source` and `dest` can't overlap, and any order of running the loop, vectorization etc. are ok ("pointer aliasing" is a keyword here). So in summary, `restrict` can make things faster, sometimes a tiny bit, sometimes a lot, depends.

For `main()` it is actually wrong. That's the "API" the OS calls to start a program and give it the command line parameters. It's declared as `int main(int, char**)`, and while some compilers allow you to change it, it won't have an effect on what the OS does. Plus - I don't see a point in it here. Even if it had a performance impact, how many command line arguments can there be?

So you are fine to patch this. Might want to use a patch instead of sed, to have context for people reading the recipes and not accidentally remove `restrict` where it might be needed correct. In this case, either is fine I guess.

You don't have to patch only on OSX. It's a fix for both - future GCC versions are not unlikely to complain (rightfully) just as CLANG is already doing.

Reply:
Many thanks for the detailed explanation @epruesse - this totally makes sense why the fix worked!

I would prefer patch as a cleaner solution as well - just the first step of the build is to "unzip" "defiant.zip" that contains "defiant.c". Since the patch is applied before "unzip defiant.zip", the build fails because "defiant.c" is not found, so `sed` was a workaround...

Currently `regions_of_interest` is not part of "defiant.zip", but if the developer adds `regions_of_interest` in "defiant-1.1.4.zip" as requested in his repo's issue, then using patch will be feasible.

Reply:
Ah - didn't think of that. True.
Let's see if we get an updated release archive. If not, feel free to merge.

Reply:
Thanks for looking at this recipe @epruesse . The developer updated the tarball, so I went ahead and merged the recipe with the proper changes.